### PR TITLE
Phase 4: Advanced Feature Engineering — Zone/Boundary/Wall-Normal Encodings (8 parallel)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
The current 24-dim input features miss several physically meaningful signals that are available in the raw data. Adding these features could significantly improve predictions, especially in regions where the model currently struggles (boundary layers, wake regions, tandem interaction zones):

1. **Boundary type one-hot encoding** (7 dims): The model currently only knows "is_surface" (binary). But boundary types 0-7 encode rich information: interior vs inlet vs outlet vs wall vs foil1 vs foil2. This is critical for tandem cases where foil1/foil2 distinction matters.

2. **Zone ID encoding** (3 dims): Mesh zone 0 (background), zone 1 (foil1 refinement), zone 2 (foil2 refinement) tell the model about local mesh resolution and proximity to foils.

3. **Wall-normal direction vectors** (2 dims): For surface nodes, compute the outward-facing normal direction from dsdf gradients. This helps the model understand flow direction relative to the surface — critical for pressure prediction.

4. **Local mesh density** (1 dim): Number of nearby nodes (proxy via dsdf magnitudes). Denser mesh regions are more important.

5. **Relative position to foil centroids** (4 dims): Distance and angle to foil 1 and foil 2 centroids. Helps the model understand spatial context in tandem configurations.

**Key insight:** The current model gets all its geometric understanding from pos (2), saf (2), dsdf (8), and is_surface (1). That's only 13 geometric dims. Adding boundary type, zone, and wall-normal features could add 12 more geometric dims, nearly doubling the geometric information available.

### Data Pipeline Changes (modify `prepare_multi.py`)

**NOTE: You have permission to modify `prepare_multi.py` for Phase 4. Do NOT modify split files.**

1. **Modify `preprocess_sample_multi()` to add new features:**

```python
def preprocess_sample_multi_v2(sample, add_boundary_onehot=False, add_zone=False, 
                                add_wall_normal=False, add_mesh_density=False,
                                add_foil_relative=False):
    """Extended preprocessing with additional physics-informed features."""
    # ... existing code up to building x ...
    
    extra_features = []
    
    if add_boundary_onehot:
        # One-hot encode boundary type (0-7)
        boundary = sample.boundary.long()
        boundary_oh = torch.zeros(n, 8, dtype=torch.float32)
        boundary_oh.scatter_(1, boundary.unsqueeze(1).clamp(0, 7), 1.0)
        extra_features.append(boundary_oh)  # +8 dims
    
    if add_zone:
        # Zone ID as features (0=background, 1=foil1-refine, 2=foil2-refine)
        zone = sample.zoneID.float() if hasattr(sample, 'zoneID') else torch.zeros(n)
        zone_oh = torch.zeros(n, 3, dtype=torch.float32)
        zone_idx = zone.long().clamp(0, 2)
        zone_oh.scatter_(1, zone_idx.unsqueeze(1), 1.0)
        extra_features.append(zone_oh)  # +3 dims
    
    if add_wall_normal:
        # Approximate wall-normal from dsdf gradient for surface nodes
        dsdf = sample.dsdf.float()  # [N, 8]
        # Use first dsdf gradient pair as surface normal proxy
        nx = dsdf[:, 0:1]  # x-component of nearest surface gradient
        ny = dsdf[:, 1:2]  # y-component
        normal_mag = (nx**2 + ny**2).sqrt().clamp(min=1e-6)
        wall_normal = torch.cat([nx / normal_mag, ny / normal_mag], dim=1)
        # Zero out for non-surface nodes
        wall_normal = wall_normal * is_surface.float().unsqueeze(1)
        extra_features.append(wall_normal)  # +2 dims
    
    if add_mesh_density:
        # Proxy for local mesh density: inverse of min dsdf magnitude
        dsdf_mag = dsdf.norm(dim=-1, keepdim=True).clamp(min=1e-6)
        mesh_density = 1.0 / dsdf_mag
        mesh_density = torch.log1p(mesh_density)  # log scale
        extra_features.append(mesh_density)  # +1 dim
    
    if add_foil_relative:
        # Distance and angle to foil centroids
        pos = sample.pos.float()  # [N, 2]
        foil1_mask = (sample.boundary == 5) | (sample.boundary == 6)
        if foil1_mask.any():
            foil1_center = pos[foil1_mask].mean(0, keepdim=True)  # [1, 2]
        else:
            foil1_center = torch.zeros(1, 2)
        
        foil2_mask = sample.boundary == 7
        if foil2_mask.any():
            foil2_center = pos[foil2_mask].mean(0, keepdim=True)
        else:
            foil2_center = foil1_center  # fallback for single-foil
        
        # Relative distance and angle
        d1 = pos - foil1_center
        d2 = pos - foil2_center
        r1 = d1.norm(dim=-1, keepdim=True).clamp(min=1e-6)
        r2 = d2.norm(dim=-1, keepdim=True).clamp(min=1e-6)
        angle1 = torch.atan2(d1[:, 1:2], d1[:, 0:1])
        angle2 = torch.atan2(d2[:, 1:2], d2[:, 0:1])
        extra_features.append(torch.cat([torch.log1p(r1), angle1, torch.log1p(r2), angle2], dim=1))  # +4 dims
    
    if extra_features:
        x = torch.cat([x] + extra_features, dim=1)
    
    return x, y, is_surface
```

2. **Update `X_DIM` dynamically** based on which features are enabled. The model's `fun_dim` must match.

3. **Add CLI flags to train.py Config:**
```python
add_boundary_onehot: bool = False
add_zone: bool = False
add_wall_normal: bool = False
add_mesh_density: bool = False
add_foil_relative: bool = False
```

4. **Update the model's `fun_dim` calculation** in train.py to account for extra features.

### GPU Assignments

| GPU | Experiment | Features added |
|-----|-----------|---------------|
| 0 | Boundary one-hot only | `--add_boundary_onehot` |
| 1 | Zone encoding only | `--add_zone` |
| 2 | Wall-normal + mesh density | `--add_wall_normal --add_mesh_density` |
| 3 | Foil-relative positions | `--add_foil_relative` |
| 4 | All features combined | `--add_boundary_onehot --add_zone --add_wall_normal --add_mesh_density --add_foil_relative` |
| 5 | Top-3 features (boundary + zone + wall-normal) | `--add_boundary_onehot --add_zone --add_wall_normal` |
| 6 | Boundary + foil-relative (tandem-focused) | `--add_boundary_onehot --add_foil_relative` |
| 7 | Baseline seed 42 | Standard baseline |

### Training Commands

```bash
# GPU 0: Boundary one-hot
CUDA_VISIBLE_DEVICES=0 python train.py --agent nezuko --wandb_name "nezuko/p4-feat-boundary" \
  --wandb_group phase4-features \
  --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp \
  --n_layers 3 --slice_num 96 --tandem_ramp \
  --domain_layernorm --domain_velhead --ema_decay 0.999 \
  --add_boundary_onehot &

# GPU 1: Zone encoding
CUDA_VISIBLE_DEVICES=1 python train.py --agent nezuko --wandb_name "nezuko/p4-feat-zone" \
  --wandb_group phase4-features \
  --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp \
  --n_layers 3 --slice_num 96 --tandem_ramp \
  --domain_layernorm --domain_velhead --ema_decay 0.999 \
  --add_zone &

# GPU 2: Wall-normal + mesh density
CUDA_VISIBLE_DEVICES=2 python train.py --agent nezuko --wandb_name "nezuko/p4-feat-wallnorm-density" \
  --wandb_group phase4-features \
  --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp \
  --n_layers 3 --slice_num 96 --tandem_ramp \
  --domain_layernorm --domain_velhead --ema_decay 0.999 \
  --add_wall_normal --add_mesh_density &

# GPU 3: Foil-relative positions
CUDA_VISIBLE_DEVICES=3 python train.py --agent nezuko --wandb_name "nezuko/p4-feat-foilrel" \
  --wandb_group phase4-features \
  --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp \
  --n_layers 3 --slice_num 96 --tandem_ramp \
  --domain_layernorm --domain_velhead --ema_decay 0.999 \
  --add_foil_relative &

# GPU 4: All features combined
CUDA_VISIBLE_DEVICES=4 python train.py --agent nezuko --wandb_name "nezuko/p4-feat-all" \
  --wandb_group phase4-features \
  --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp \
  --n_layers 3 --slice_num 96 --tandem_ramp \
  --domain_layernorm --domain_velhead --ema_decay 0.999 \
  --add_boundary_onehot --add_zone --add_wall_normal --add_mesh_density --add_foil_relative &

# GPU 5: Top-3 features
CUDA_VISIBLE_DEVICES=5 python train.py --agent nezuko --wandb_name "nezuko/p4-feat-top3" \
  --wandb_group phase4-features \
  --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp \
  --n_layers 3 --slice_num 96 --tandem_ramp \
  --domain_layernorm --domain_velhead --ema_decay 0.999 \
  --add_boundary_onehot --add_zone --add_wall_normal &

# GPU 6: Tandem-focused features
CUDA_VISIBLE_DEVICES=6 python train.py --agent nezuko --wandb_name "nezuko/p4-feat-tandem" \
  --wandb_group phase4-features \
  --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp \
  --n_layers 3 --slice_num 96 --tandem_ramp \
  --domain_layernorm --domain_velhead --ema_decay 0.999 \
  --add_boundary_onehot --add_foil_relative &

# GPU 7: Baseline
CUDA_VISIBLE_DEVICES=7 python train.py --agent nezuko --wandb_name "nezuko/p4-baseline-seed42" \
  --wandb_group phase4-baseline-seeds \
  --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp \
  --n_layers 3 --slice_num 96 --tandem_ramp \
  --domain_layernorm --domain_velhead --ema_decay 0.999 \
  --seed 42 &
wait
```

## Baseline
| Metric | Value |
|--------|-------|
| val/loss | 0.3994 |
| p_in | 13.0 |
| p_oodc | 8.7 |
| p_tan | 33.2 |
| p_re | 24.6 |

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT nezuko:

## Results

### Experiment Matrix (all 8 GPUs, 180 min timeout)

| GPU | Experiment | Extra Dims | Best Epoch | val/loss | p_in | p_oodc | p_tan | p_re |
|-----|-----------|-----------|------------|----------|------|--------|-------|------|
| -- | **BASELINE (ref)** | 0 | -- | **0.3994** | **13.0** | **8.7** | **33.2** | **24.6** |
| 7 | Baseline seed 42 | 0 | 160 | 0.4088 | 13.5 | 8.4 | 33.8 | 24.6 |
| 0 | Boundary one-hot | +8 | 158 | 0.4082 | **12.9** | 8.9 | 34.2 | 24.7 |
| 1 | Zone encoding | +3 | 159 | 0.4014 | 13.8 | **8.3** | 33.2 | **24.5** |
| 2 | Wall-normal + density | +3 | 159 | 0.4119 | 13.9 | 9.0 | 33.3 | 24.7 |
| 3 | Foil-relative | +4 | 157 | 0.4050 | 13.2 | 8.7 | 33.2 | 24.8 |
| 4 | All features | +18 | 158 | 0.4074 | 14.0 | 8.6 | 33.6 | 24.7 |
| 5 | Top-3 (boundary+zone+wall) | +13 | 157 | 0.4052 | 13.4 | 8.5 | 33.8 | 24.8 |
| 6 | Tandem-focused (boundary+foil) | +12 | 158 | 0.4046 | 13.5 | 8.4 | 33.3 | 24.7 |

### W&B Runs

| Experiment | Run |
|---|---|
| GPU 0: Boundary one-hot | [1ywopkf7](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/1ywopkf7) |
| GPU 1: Zone encoding | [mayl8s9g](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/mayl8s9g) |
| GPU 2: Wall-normal + density | [dmnofw6y](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/dmnofw6y) |
| GPU 3: Foil-relative | [lth83lr7](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/lth83lr7) |
| GPU 4: All features | [c4ow46pi](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/c4ow46pi) |
| GPU 5: Top-3 | [9l88mesu](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/9l88mesu) |
| GPU 6: Tandem-focused | [vfumx7qk](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/vfumx7qk) |
| GPU 7: Baseline seed 42 | [z1mit2lo](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/z1mit2lo) |

### Commands Used

All experiments used the baseline config plus feature flags:
```bash
# Base command (all experiments share this)
python train.py --agent nezuko --wandb_name "nezuko/p4-feat-<desc>" \
  --wandb_group phase4-features \
  --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp \
  --n_layers 3 --slice_num 96 --tandem_ramp \
  --domain_layernorm --domain_velhead --ema_decay 0.999 \
  [--add_boundary_onehot] [--add_zone] [--add_wall_normal] [--add_mesh_density] [--add_foil_relative]
```

Peak VRAM: ~42 GB per experiment (96 GB GPUs).

### What Happened

**No experiment clearly beats the baseline on val/loss.** The results are mixed across individual metrics:

**Zone encoding (GPU 1)** was the strongest variant overall — it achieved the best p_oodc (8.3 vs 8.7 baseline) and p_re (24.5 vs 24.6), while matching p_tan (33.2). However, it traded off worse p_in (13.8 vs 13.0).

**Boundary one-hot (GPU 0)** was the only variant that improved p_in (12.9 vs 13.0), suggesting boundary type information helps in-distribution predictions. But it hurt p_tan (34.2 vs 33.2).

**Wall-normal + mesh density (GPU 2)** performed worst overall, suggesting these derived geometric features add noise rather than signal. The model likely already captures this information through dsdf.

**Combined approaches (GPU 4-6)** showed no synergy — stacking more features didn't improve over individual ones. This suggests the extra dimensions add noise that offsets any benefit from the additional information.

**Baseline seed 42 (GPU 7)** shows natural variance: val/loss=0.4088 vs reference 0.3994, confirming that differences of ~0.01 in val/loss are within seed noise.

### Analysis

The core insight is that the current 24-dim feature set already captures most of the physics the model needs. The additional features provide marginal or no improvement because:

1. **Boundary type**: Most of the boundary info is already encoded via `is_surface` and the dsdf gradients. The one-hot adds redundant info for most node types.
2. **Zone ID**: Provides a small signal for OOD generalization (zone structure hints at mesh topology), but the model can largely infer this from dsdf patterns.
3. **Wall-normal/mesh density**: Fully derivable from existing dsdf features — the model is already learning these implicitly.
4. **Foil-relative positions**: The model already has pos + dsdf which encodes relative geometry.

### Suggested Follow-ups

1. **Zone encoding is the most promising** — the OOD improvements warrant further investigation. Could try combining zone with a reduced model width to test if the signal persists without extra capacity.
2. **Feature selection via learned gates** — instead of fixed one-hot features, could add a learnable attention gate on the extra features so the model can dynamically weight their importance.
3. **The visualize() function has a pre-existing signature mismatch bug** — `train.py` calls `visualize(samples, out_dir=...)` but `data/utils.py` expects `visualize(model, val_ds, stats, device, ...)`. This prevents flow field plots from being generated. Not related to this PR but worth fixing.
